src/components/macro/macro.component.ts

```diff
@@ -1,12 +1,132 @@
-import { Component } from '@angular/core';
+import { AfterViewInit, Component, OnDestroy, OnInit, QueryList, ViewChildren } from '@angular/core';
+import { ActivatedRoute } from '@angular/router';
+import { cloneDeep as _cloneDeep } from 'lodash';
+import { DragulaService } from 'ng2-dragula/ng2-dragula';
+import { Macro } from '../../config-serializer/config-items/Macro';
+import { MacroItemComponent } from './macro-item/macro-item.component';
+import { Subscription } from 'rxjs/Subscription';
+import { TextMacroAction } from '../../config-serializer/config-items/macro-action/TextMacroAction';
+import { UhkConfigurationService } from '../../services/uhk-configuration.service';
 
 @Component({
     selector: 'macro',
     template: require('./macro.component.html'),
-    styles: [require('./macro.component.scss')]
+    styles: [require('./macro.component.scss')],
+    viewProviders: [DragulaService]
 })
-export class MacroComponent {
+export class MacroComponent implements OnInit, OnDestroy, AfterViewInit {
+    @ViewChildren(MacroItemComponent) macroItems: QueryList<MacroItemComponent>;
 
-    constructor() {
+    private macro: Macro;
+
+    private sub: Subscription;
+    private macroItemsSub: Subscription;
+    private addedNewAction: boolean = false;
+    private hasChanges: boolean = false;
+    private dragEnabled: boolean;
+
+    constructor(
+        private uhkConfigurationService: UhkConfigurationService,
+        private route: ActivatedRoute,
+        private dragulaService: DragulaService
+    ) {
+        /* tslint:disable:no-unused-variable: Used by Dragula. */
+        dragulaService.setOptions('macroActions', {
+            moves: function (el: any, container: any, handle: any) {
+                return handle.className.indexOf('action--movable') !== -1;
+            }
+        });
+        /* tslint:enable:no-unused-variable */
+    }
+
+    ngOnInit() {
+        this.sub = this.route.params.subscribe((params: any) => {
+            const id: number = params.id;
+            this.macro = this.getMacro(id);
+            this.dragEnabled = true;
+            this.hasChanges = false;
+       });
+    }
+
+    ngAfterViewInit() {
+       this.macroItemsSub = this.macroItems.changes.subscribe((data: any) => {
+           if (this.addedNewAction) {
+               // Open editor for newly added action
+               // Rather cludge way to do this, basically macroItems have to be updated before the editor can be opened
+               setTimeout(() => {
+                   let newAction = data.last;
+                   newAction.title = 'New macro action';
+                   newAction.actionEditor.toggleEnabled(true);
+                   this.hideOtherActionEditors(data.length - 1);
+                   this.addedNewAction = false;
+               });
+           }
+       });
+    }
+
+    getMacro(id: number) {
```

Comment:
This has no bearing on #95 really. It only operates on the MacroItemComponent elements on the page. Basically this is needed for opening the editor for a new macro item. So when you click Add new action, a new MacroAction is added to the array of MacroActions, then this causes a new MacroItemComponent and the MacroActionEditor nested within it to be created. To open the editor after adding a new action, we need to wait for Angular to create the MacroItemComponents to the DOM. The setTimeout is to avoid the additional check done in development mode.